Firestore/core/src/model/target_index_matcher.cc

```diff
@@ -0,0 +1,150 @@
+
+#include "Firestore/core/src/model/target_index_matcher.h"
+
+#include "Firestore/core/src/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+using core::FieldFilter;
+using core::Filter;
+using core::OrderBy;
+using core::Target;
+
+TargetIndexMatcher::TargetIndexMatcher(const core::Target& target) {
+  collection_id_ = target.collection_group() != nullptr
+                       ? *target.collection_group()
+                       : target.path().last_segment();
+  order_bys_ = target.order_bys();
+  inequality_filter_ = absl::nullopt;
+
+  for (const Filter& filter : target.filters()) {
+    FieldFilter field_filter(filter);
+    if (field_filter.IsInequality()) {
+      HARD_ASSERT(!inequality_filter_.has_value() ||
+                      inequality_filter_->field() == field_filter.field(),
+                  "Only a single inequality is supported");
+      inequality_filter_ = field_filter;
+    } else {
+      equality_filters_.push_back(field_filter);
+    }
+  }
+}
+
+bool TargetIndexMatcher::ServedByIndex(const model::FieldIndex& index) {
+  HARD_ASSERT(index.collection_group() == collection_id_,
+              "Collection IDs do not match");
+
+  // If there is an array element, find a matching filter.
+  const auto& array_segment = index.GetArraySegment();
+  if (array_segment.has_value() &&
+      !HasMatchingEqualityFilter(array_segment.value())) {
+    return false;
+  }
+
+  std::vector<Segment> segments = index.GetDirectionalSegments();
+  size_t segment_index = 0;
+  // Process all equalities first. Equalities can appear out of order.
+  for (; segment_index < segments.size(); ++segment_index) {
+    // We attempt to greedily match all segments to equality filters. If a
+    // filter matches an index segment, we can mark the segment as used. Since
+    // it is not possible to use the same field path in both an equality and
+    // inequality/oderBy clause, we do not have to consider the possibility that
+    // a matching equality segment should instead be used to map to an
+    // inequality filter or orderBy clause.
+    if (!HasMatchingEqualityFilter(segments[segment_index])) {
+      // If we cannot find a matching filter, we need to verify whether the
+      // remaining segments map to the target's inequality and its orderBy
+      // clauses.
+      break;
+    }
+  }
+
+  // If we already have processed all segments, all segments are used to serve
+  // the equality filters and we do not need to map any segments to the target's
+  // inequality and orderBy clauses.
+  if (segment_index == segments.size()) {
+    return true;
+  }
+
+  auto order_by_iter = order_bys_.begin();
+
+  // If there is an inequality filter, the next segment must match both the
+  // filter and the first OrderBy clause.
+  if (inequality_filter_.has_value()) {
+    if (!MatchesFilter(inequality_filter_, segments[segment_index]) ||
+        !MatchesOrderBy(*(order_by_iter++), segments[segment_index])) {
```

Comment:
It's probably better to explicitly increment the iterator rather than hide it inline (here and below).

Reply:
We should also maybe add a comment to line 86 that says that the orderBy is guaranteed to have at least one element.